paddle/operators/adadelta_op.h

```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename Place, typename T>
+class AdadeltaOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto param_out = ctx.Output<framework::Tensor>("ParamOut");
+    auto avg_squared_grad_out =
+        ctx.Output<framework::Tensor>("AvgSquaredGradOut");
+    auto avg_squared_update_out =
+        ctx.Output<framework::Tensor>("AvgSquaredUpdateOut");
+
+    param_out->mutable_data<T>(ctx.GetPlace());
+    avg_squared_grad_out->mutable_data<T>(ctx.GetPlace());
+    avg_squared_update_out->mutable_data<T>(ctx.GetPlace());
+
+    float rho = ctx.Attr<float>("rho");
+    float epsilon = ctx.Attr<float>("epsilon");
+
+    auto p = framework::EigenVector<T>::Flatten(
+        *ctx.Input<framework::Tensor>("Param"));
+    auto g = framework::EigenVector<T>::Flatten(
+        *ctx.Input<framework::Tensor>("Grad"));
+    // Squared gradient accumulator
+    auto g_acc = framework::EigenVector<T>::Flatten(
+        *ctx.Input<framework::Tensor>("AvgSquaredGrad"));
+    // Squared updates accumulator
+    auto u_acc = framework::EigenVector<T>::Flatten(
+        *ctx.Input<framework::Tensor>("AvgSquaredUpdate"));
+    auto p_out = framework::EigenVector<T>::Flatten(*param_out);
+    auto g_acc_out = framework::EigenVector<T>::Flatten(*avg_squared_grad_out);
+    auto u_acc_out =
+        framework::EigenVector<T>::Flatten(*avg_squared_update_out);
+    auto place = ctx.GetEigenDevice<Place>();
+
+    g_acc_out.device(place) = rho * g_acc + (1 - rho) * g.square();
```

Comment:
maybe we can name `g_acc_out` as `avg_squared_grad_eigen` to be consistent with the formula written in the DOC, it will be better to read and understand

Reply:
Sure, I will make the change

Reply:
since `avg_squared_grad_eigen` is a litter too long, we can use `avg_squared_grad` here and  change the formal value `avg_squared_grad`  get from tensor to `avg_squared_grad_t` or something like this